### ISSUE TITLE
flat displays inconsistent names and search/filter won't match flattened subs

### ISSUE THREAD

**Issue by ri-aje** [ranger/ranger#2044]
#### Runtime Environment

- Operating system and version: 
- Terminal emulator and version: 
- Python version: 
- Ranger version/commit: 
ranger version: ranger-master v1.9.3-27-gd17e532a
Python version: 3.7.1 (default, Sep 21 2019, 10:53:58) [GCC 7.4.0]
Locale: en_US.UTF-8
- Locale: 

#### Current Behavior
1. `flat` doesn't consistently display the full prefix for a folder/file. sometimes it displays the full relative path, and sometimes just the plain leaf folder name. flat doesn't even do this consistently, it mixes both styles in one view. and the exact display changes as my cursor moves across the flattened entities. lots of inconsistency in the display. might be the same issue reported in https://github.com/ranger/ranger/issues/1085, that one looks like getting no attention so far (2+ years).

1. `search`, `/` and `filter` doesn't seem to honor any `flat` in effect, i.e., they don't operate on all the entities in the current flat view, instead they seem to operate on the unflattened view which is invisible to the user when a `flat` is in effect. confusing and unexpected.

#### Expected Behavior
1. `flat` to consistently show all full relative prefix for all sub folders/files and show them all the time consistently with or without cursor movement.

1. `search`, `/` and `filter` should operate on the current visible view to the user. if there is a flat in effect and a bunch of sub entities that are otherwise invisible are currently visible, then those commands should consider them, otherwise there is a clear break between what users see and what ranger sees, resulting in counter intuitive experience.

#### Context

partially dysfunctional flat. some tasks cannot be easily completed. for example, I cannot `flat 1` in a folder full of subfolders, then `filter` out two target folders then quickly select everything underneath them, and operate (e.g., cut them) on those sub entities only under my target folders. 

#### Possible Solutions
1. looks like a bug, so just fix the behavior up to expectation.

1. make the operation targets of `search`, `/` and `filter` consistent with what is currently displayed.

#### Steps to reproduce
1. `mkdir -p test/a test/b`
```
.            
└── test/ 
    ├── a/
    └── b/
```
2. launch ranger

3. `:flat -1` to show everything, I got this.
![image](https://user-images.githubusercontent.com/321429/87213205-39670a00-c2d8-11ea-988d-a4b048615e03.png)
note `b` is missing the `test/` prefix.

4. moving my cursor one up, I got this.
![image](https://user-images.githubusercontent.com/321429/87213224-58fe3280-c2d8-11ea-8490-68ad55cb5109.png)
now everything looks as expected except the order, but only if my cursor stays here.

5. moving one more up, I got this.
![image](https://user-images.githubusercontent.com/321429/87213239-729f7a00-c2d8-11ea-9871-217e2c172aa4.png)
note the inconsistency isn't even consistent, since `a` and `b` both miss their prefix now and it was just `b` earlier.

6. now `filter test`, I expect to see all stay visible since they should all match. the sub entities should match due to having the `test/` prefix from `flat`. but no, I only see `test` now.

7. same problem with `search` and `/`. if I search for `test`, cursor jumps to the `test` folder directly and pressing `n` or `N` goes nowhere next, skipping `test/a` and `test/b` completely.

8. about the listing order, I would expect 
```
test
test/a
test/b
```
like what `find . -type d` would give me. it makes sense to list a folder first then all the entities in it, just like see a folder first, then only after cd into it, I will see all stuff in it. 

9. I observed the same issue when `a` and `b` are files.

#### Traceback

```

```

**Comment by toonn** [comment]
Closing this in favor of #1085.

The reason filtering for "test" doesn't work is because that filters by basename, not displayed name. The newer `filter_stack` "name" filter (`.n`) applies a regex to the file's relative path so should work in this case.

**Comment by ri-aje** [comment]
> Closing this in favor of #1085.
> 
> The reason filtering for "test" doesn't work is because that filters by basename, not displayed name. The newer `filter_stack` "name" filter (`.n`) applies a regex to the file's relative path so should work in this case.

indeed the filter stack works. thanks.

@toonn if you decide to fold this into https://github.com/ranger/ranger/issues/1085, can you make sure to fix the display order as well. I don't see https://github.com/ranger/ranger/issues/1085 asked for that.

and how about search? will it get improved to target display names? filter is kind of taken care of by the filter stack, does search have a similar alternative?

**Comment by toonn** [comment]
Usually I go through linked PRs and issues but if you comment those things on that issue I definitely won't miss it.

**Comment by ri-aje** [comment]
> Usually I go through linked PRs and issues but if you comment those things on that issue I definitely won't miss it.

I see this issue is linked in https://github.com/ranger/ranger/issues/1085, so I guess you will see my comments then.

> 
> Oh, and if you want to get involved this actually sounds like a good issue to work on. Not too much work I think but a pretty visible change.

feel free to point me a place to start. and btw, don't count on me as I don't know how much resource I can put in here. but if I can get it before you, I will share.

**Comment by toonn** [comment]
I'd start by taking a look at the flat and scout commands, probably in `ranger/core/actions.py`.

**Comment by ri-aje** [comment]
> I'd start by taking a look at the flat and scout commands, probably in `ranger/core/actions.py`.

I see `load_bit_by_bit` does the heavy lifting, managed by some external scheduler for non-blocking I guess. the directory object gets filled with both 
```
                self.filenames = filenames
                self.files_all = files
```
which one of them is used for display?

there is a `self.sort()` call at the end, which only sorts on `files_all`, is sorting `filenames` the missing piece?

also, how do I do printf debugging with ranger? `print` anything just messes up my screen, and hard to tell what's my prints.

I need to do some prints to take a look at the internals, and see where things go off the trail.

**Comment by toonn** [comment]
`files_all` contains File objects and is what's used for everything afaik. You can print to ranger's log using `self.fm.notify("message", bad=True if error)`.

**Comment by ri-aje** [comment]
> `files_all` contains File objects and is what's used for everything afaik. You can print to ranger's log using `self.fm.notify("message", bad=True if error)`.

how do I access ranger log (ideally as plain file)? if I print a large `files_all` (several thousands items), the status bar only shows it a short while and only shows one line. pressing `W` takes a while to load, and I cannot see the entire message, only what's fit in my screen (obviously not big enough to show everything), and I cannot copy the stuff out for further formatting either, the area denies any mouse selection. these limitations make it practically impossible to debug this way. with https://github.com/ri-aje/ranger/commit/e13536cd38ac86ea66f1fddadf3a9cfa15a1630f, it seems to have improved and I cannot see those inconsistencies easily on some small test dirs now, but only notice them on some larger ones.

how do you debug ranger?

**Comment by ri-aje** [comment]
after quite some debugging, here is what I found.
1. the inconsistencies in missing prefix dirnames are largely gone in https://github.com/ri-aje/ranger/commit/e13536cd38ac86ea66f1fddadf3a9cfa15a1630f, looks like they got fixed for whatever reason.
1. the ordering is still undesirable with folder listed after its children and here are some inconsistencies as well. some times parent dir got listed last and sometimes got listed first. the root cause is this line of sorting https://github.com/ri-aje/ranger/blob/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae/ranger/container/directory.py#L535, which gives unexpected result for `sort_naturally_icase` while it works as expected for `sort_by_basename_icase`. `basename_natural_lower` (https://github.com/ri-aje/ranger/blob/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae/ranger/container/fsobject.py#L168) looks good to me. my tests are all lowercase English chars, so the `icase` here shouldn't matter and `basename_natural_lower` should always go into the `except` clause to just grab the split strings. I double checked this piece of code by splitting it out into a tiny python program and it gives me the expected sorting there. inside `fsobject`, however, it has to behave unexpectedly and I am not sure why. what's even more confusing is even when I just directly `return self.relative_path_lower` in `basename_natural_lower`, it still doesn't work. totally unexpected since that's just `sort_by_basename_icase` which works perfectly.
1. along the way, I did fix the `filter` command to work on relative path. https://github.com/ri-aje/ranger/commit/6848db204cb6d02fb3f29414081261c63b6dc6ea makes `filter` more intuitive in flattened view.
1. and I also fixed the `search` command the same way in https://github.com/ri-aje/ranger/commit/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae. let me know if they are good to merge into upstream.

**Comment by ri-aje** [comment]
> @toonn after quite some debugging, here is what I found.
> 
> 1. the inconsistencies in missing prefix dirnames are largely gone in [ri-aje@e13536c](https://github.com/ri-aje/ranger/commit/e13536cd38ac86ea66f1fddadf3a9cfa15a1630f), looks like they got fixed for whatever reason.
> 2. the ordering is still undesirable with folder listed after its children and here are some inconsistencies as well. some times parent dir got listed last and sometimes got listed first. the root cause is this line of sorting https://github.com/ri-aje/ranger/blob/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae/ranger/container/directory.py#L535, which gives unexpected result for `sort_naturally_icase` while it works as expected for `sort_by_basename_icase`. however, `basename_natural_lower` (https://github.com/ri-aje/ranger/blob/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae/ranger/container/fsobject.py#L168) looks good to me. my tests are all lowercase English chars, so the `icase` here shouldn't matter and `basename_natural_lower` should always go into the `except` clause to just grab the split strings. I double checked this piece of code by splitting it out into a tiny python program and it gives me the expected sorting there. inside `fsobject`, however, it has to behave unexpectedly and I am not sure why. what's even more confusing is even when I just directly `return self.relative_path_lower` in `basename_natural_lower`, it still doesn't work. totally unexpected since that's just `sort_by_basename_icase` which works perfectly.
> 3. along the way, I did fix the `filter` command to work on relative path. [ri-aje@6848db2](https://github.com/ri-aje/ranger/commit/6848db204cb6d02fb3f29414081261c63b6dc6ea) makes `filter` more intuitive in flattened view.
> 4. and I also fixed the `search` command the same way in [ri-aje@37bc46a](https://github.com/ri-aje/ranger/commit/37bc46aade1d936216f00a38c75d2fd4c1c8f3ae). let me know if they are good to merge into upstream.
> 
> oh, btw, `filter` and `filter_stack` aren't the same thing. the latter is much more comprehensive and complicated to reason (with a stack structure) and it also doesn't filter as I type, so it's hard to figure out if my filter does what I want, whereas `filter` gives me instant feedback. they are both useful, but `filter_stack` is geared towards more heavy-duty jobs.

ranger is really acting up on me. right after I finished the previous comment and tried again, the same missing prefix inconsistency showed up again, on the exact test that has been working fine for the past few hours, so I have to take back the above first item. with this come and go nature, this now feels like some data race issue under the hood.
